unified_ecommerce/test_utils.py

```diff
@@ -138,3 +223,211 @@ class PickleableMock(Mock):
+        """
+        Test that the viewset can retrieve an object.
+
+        Args:
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+        """
+        instance = self.factory_class()
+        instance_qs = self.queryset.filter(pk=instance.pk)
+
+        response = self._test_retrieval(
+            user_client if isLoggedIn else client,
+            self.object_url.format(instance.pk),
+            "object_url",
+            test_non_authenticated=not isLoggedIn,
+        )
+
+        if isLoggedIn and instance.is_active is None:
+            dj_serializer = queryset_to_json(instance_qs)
+            assert_json_equal(
+                *make_timestamps_matchable([response.data, dj_serializer])
+            )
+
+        if isLoggedIn and instance.is_active is not None:
+            # Deleted items should return a 404. (is_active is None if it's active;
+            # otherwise it has a timestamp of when it was deleted.)
+            assert response.status_code == 404
+
+    def test_update(self, update_data, isLoggedIn, client, user_client):
+        """
+        Test that the viewset can update an object.
+
+        Args:
+        - update_data (dict): the data to use for the update
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+
+        Returns:
+        - tuple of (instance, response): the instance that was updated and the response
+        """
+        instance = self.factory_class()
+
+        use_client = user_client if isLoggedIn else client
+        response = use_client.patch(
+            self.object_url.format(instance.pk), data=update_data
+        )
+
+        if not isLoggedIn:
+            assert response.status_code == 403
+        return (instance, response)
+
+    def test_delete(self, isLoggedIn, client, user_client):
+        """
+        Test that the viewset can delete an object. Note that this will actually test
+        deletion.
+
+        Args:
+        - isLoggedIn (bool): whether or not the client is logged in
+        - client (APIClient): the client to use for non-logged in requests
+        - user_client (APIClient): the client to use for logged in requests
+
+        Returns:
+        - tuple of (instance, response): the instance that was deleted and the response
+        """
+        instance = self.factory_class()
+
+        use_client = user_client if isLoggedIn else client
+        response = use_client.delete(self.object_url.format(instance.pk))
+
+        assert response.status_code == 403 if not isLoggedIn else 204
+
+        if isLoggedIn:
+            assert self.queryset.filter(pk=instance.pk).count() == 0
```

Comment:
Might be a little more robust and comprehensive to take a count of records before the test and then if they're logged in the expected count is `old_count - 1` otherwise `old_count`.